File: sdk/ruby/lib/anyvali/schemas/tuple_schema.rb
# frozen_string_literal: true

module AnyVali
  class TupleSchema < Schema
    attr_reader :element_schemas

    def initialize(elements:, **kwargs)
      @element_schemas = elements.freeze
      super(kind: "tuple", **kwargs)
    end

    def to_node
      node = super
      node["elements"] = @element_schemas.map(&:to_node)
      node
    end

    protected

    def validate(value, path, issues, context)
      unless value.is_a?(Array)
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: "tuple",
          received: Schema.type_name(value)
        )
        return
      end

      expected_len = @element_schemas.length
      if value.length < expected_len
        issues << ValidationIssue.new(
          code: IssueCodes::TOO_SMALL,
          path: path,
          expected: expected_len.to_s,
          received: value.length.to_s
        )
        return
      end

      if value.length > expected_len
        issues << ValidationIssue.new(
          code: IssueCodes::TOO_LARGE,
          path: path,
          expected: expected_len.to_s,
          received: value.length.to_s
        )
        return
      end

      @element_schemas.each_with_index do |schema, i|
        result = schema.safe_parse(value[i], path: path + [i], context: context)
        issues.concat(result.issues) if result.failure?
      end
    end

    def dup_with(**overrides)
      elements = overrides.delete(:elements) || @element_schemas
      attrs = {
        elements: elements,
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators
      }.merge(overrides)
      self.class.new(**attrs)
    end
  end
end

File: sdk/csharp/src/AnyVali/Schemas/TupleSchema.cs
namespace AnyVali.Schemas;

public sealed class TupleSchema : Schema
{
    private readonly List<Schema> _items;

    public TupleSchema(IEnumerable<Schema> items)
    {
        _items = items.ToList();
    }

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        if (input is not List<object?> arr)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.InvalidType,
                Message = $"Expected tuple, received {DescribeType(input)}",
                Path = ctx.ClonePath(),
                Expected = "tuple",
                Received = DescribeType(input),
            });
            return null;
        }

        if (arr.Count < _items.Count)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooSmall,
                Message = $"Tuple must have exactly {_items.Count} element(s), received {arr.Count}",
                Path = ctx.ClonePath(),
                Expected = _items.Count.ToString(),
                Received = arr.Count.ToString(),
            });
            return null;
        }

        if (arr.Count > _items.Count)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooLarge,
                Message = $"Tuple must have exactly {_items.Count} element(s), received {arr.Count}",
                Path = ctx.ClonePath(),
                Expected = _items.Count.ToString(),
                Received = arr.Count.ToString(),
            });
            return null;
        }

        var result = new List<object?>();
        for (var i = 0; i < _items.Count; i++)
        {
            ctx.PushPath(i);
            var val = _items[i].RunPipeline(arr[i], ctx);
            result.Add(val);
            ctx.PopPath();
        }
        return result;
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?>
        {
            ["kind"] = "tuple",
            ["elements"] = _items.Select(s => (object?)s.ToNode()).ToList(),
        };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new TupleSchema(_items)
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}
